Refactor the integration testing code

YehorBoiar opened the pull request:
This PR is created to address the issue #523

YehorBoiar commented:
@ozgurakgun Please review

niklasdewally commented:
Thank you for improving the docs as part of this :)

niklasdewally commented on conjure_oxide/tests/generated_tests.rs:
```diff
@@ -148,156 +148,195 @@ fn integration_test_inner(
-    // Stage 0: Compare the two methods of parsing
-    // skip if the field is set to true
-    // do not skip if it is unset, or if it is explicitly set to false
+    // Stage 0: Compare native parser with expected (unless skipped)
+    let mut model_native = None;
```
```suggestion
    // Stage 0: Run native parser (unless skipped)
    let mut model_native = None;
```

Optional: we aren't actually doing the comparison here, but later

niklasdewally commented on conjure_oxide/tests/generated_tests.rs:
```diff
@@ -148,156 +148,195 @@ fn integration_test_inner(
     }
 
-    // Stage 1: Read the essence file and check that the model is parsed correctly
+    // Stage 1: Parse the model using the normal parser
```
```suggestion
    // Stage 1: Parse the model using the Conjure-JSON parser
```

Pedantic

niklasdewally commented on conjure_oxide/tests/generated_tests.rs:
```diff
@@ -148,156 +148,195 @@ fn integration_test_inner(
+        save_stats_json(context, path, essence_base)?;
+        return Ok(());
+    }
 
-    // test solutions against conjure before writing
+    // If ACCEPT = true, we skip intermediate assertions and only check conjure solutions
     if accept {
-        let mut conjure_solutions: Vec<BTreeMap<Name, Literal>> =
+        // Check solutions against Conjure
```
Pedantic: if we return at the end of the !accept block, we don't need a `if accept {` here, and can get rid of a level of nesting.

(This had me looking to see if there was any code for both at the end, which there doesn't seem to be)

niklasdewally reviewed (approved):
LGTM, aside from some minor nitpicks!

ozgurakgun commented:
I am not sure of this one yet @YehorBoiar - I feel we can tidy this up a bit more, but I need to spend a bit more time formulating my thoughts. Thanks for the work so far!

niklasdewally commented:
@ozgurakgun @YehorBoiar 

Maybe we need to think deeper about how we customise each stage of the tester

Currently we have the following steps

1. Parse
2. Rewrite
3. Solve

(thanks Yehor for documenting this)

We could think about all the other stuff the tester does as stages too: 

1a. Parse
1b. Check against other parser(s)
3a. Rewrite
3c. Check model properties (extra_asserts)
4. Solve
5. Check solutions against Conjure

These steps fall into two categories:

* Optional steps, enabled by env var / configuration value: 1b,3c,5
* Required steps with multiple concrete implementations: e.g. native parser vs json parser, naive rewriter vs optimising one.

The advantage of this would be configuration. For each optional step, we would parse a `config.toml` value `enable_<step>`. For each required step, we would keep a list of possible implementations, which could be changed in config by `<step>_impl`.

ozgurakgun commented:
(1) The steps having a similar shape to each other, maybe a step abstraction that takes in a function and implements the generate/check-against/accept logic once is closer to what I am thinking.

(2) We can also think about a global properties/assertions/warnings structure. That is enabled/disabled depending on what we are looking for in a particular run. An example is the `prop_` we added in the naive rewriter. A test run or an executable run of oxide should be able ro enable/disable different properties at different levels. Consider it an error/warning/trace etc.

I have been looking into the existing tracing libraries for (2) but didn't manage to find one that does everything I think we need. I think we need named properties, registered centrally (maybe through codegen) and when at call site we decide how to interpret each property. (1) is this PR, (2) is a separate but related one...

niklasdewally commented:
> (2) We can also think about a global properties/assertions/warnings structure. That is enabled/disabled depending on what we are looking for in a particular run. An example is the `prop_` we added in the naive rewriter. A test run or an executable run of oxide should be able ro enable/disable different properties at different levels. Consider it an error/warning/trace etc.

To clarify, these need to be dynamic, not static? If build time is fine, feature flags and simple declarative macros could take us most the way there...

Can we simply store these flags in the `Context`? `Context` already acts as a global variable store,accessible by the entire program which is initialised by the caller of Conjure Oxide API (be that the tester, CLI, user crates): this sounds like it fits the requirements above?
 

As we have serde, it would be easy to set these flags through a `toml` file instead of the command line.  And we already read toml in the for the tests anyways, so we could do:

```toml
# config.toml
[test]
# test settings here
rewriter=  "naive"

[debug_flags]
# conjure oxide debug settings here
debug_check_only_one_applicable_rule= true
```

For example, we could parse the debug_flags section into a `DebugFlags` struct which is part of our `Context`.

niklasdewally commented:
Moving our existing properties to the context and making it serialisable from the test toml should be possible to do immediately without too much effort.

 Adding macros / a nicer API on top of this would hopefully then just go ontop of our existing stuff and not require another large scale refactor.

ozgurakgun commented:
dynamic as in runtime instead of compile time, yes.

I expect we won't change this during a run, but I guess it should be possible. A kind of debugging option where we increase the message levels during a run could be useful.

putting them in context is fine.

YehorBoiar commented:
> * naive rewriter vs optimising one

We have 2 rewriters @niklasdewally?

niklasdewally commented:
@YehorBoiar see https://github.com/conjure-cp/conjure-oxide/pull/528.

I thought I pinged you in this last semester- apologies if you didn't get it!

YehorBoiar commented:
There are a lot of thigs that we want to change in generated_tests. I decided to break down the implementation of those changes into steps. I want you @ozgurakgun @niklasdewally to check if those would cover everything we want to change.

### 1. Break our general stages 1,2,3 into substages. 
```
1a. Parse (Convert the Essence model into an internal format.)
1b. Check against other parser(s) (Compare output from different parsing methods (e.g., native parser vs JSON parser).)
2a. Rewrite (Apply transformations to the parsed model (e.g., simplifying expressions, normalizing).)
2b. Check model properties (extra_asserts) (Verify additional model properties (e.g., ensure vector operators are evaluated).)
3a. Solve (Run the model through the solver to find solutions.)
3b. Check solutions against Conjure (Compare solutions with expected output from the Conjure solver.)
```
There are required stages: 1a, 2a, 3a - they have multiple implementations to them. We want to choose the implementation in `config.toml` e.g.
```toml
parser_impl = "native"
rewriter_impl = "optimizing"
```
And non-required stages we would want to choose by writing `enable_<step>` e.g.
```toml
enable_1b = true
enable_3c = true
enable_4b = false
```
All I need to do here is perhaps add more variables to `TestConfig`
```rust
struct TestConfig {
    extra_rewriter_asserts: Option<Vec<String>>,
//    skip_native_parser: Option<bool>,
//    use_naive_rewriter: Option<bool>,
    enable: Option<Vec<String>> // Perhaps we can just specify all optional stages in a vector of string instead of doing `enable_<step>`? @niklasdewally 
    impl: Option<Vec<String>> // Same here?  
// I think we wouldn't need skip_native_parser and use_naive_rewriter once we introduce `enable` and `impl` variables
}
```
As suggested by Nick, we will iterate through the enabled variables, executing only those steps, throwing errors if they fail, and rewriting if accept=true.

### 2. Step Abstraction

After ensuring that implenetation of step 1 works - we can make an abstraction for those 3 main steps. Now it has the following structure:
1. Generate something (parse model, rewrite model, solve).
2. Check output against expected results (compare parsed model, rewritten model, or solutions).
3. Accept new results if ACCEPT=true (update expected files).
We can create a function that would take a function in as a parameter - use that function to generate result we need, check against expected results, and then check if we watn to rewrite (accept=true)

### 3. Global Properties/Assertions/Warnings

I don't understand what do we want here... @ozgurakgun 
```
(2) We can also think about a global properties/assertions/warnings structure. That is enabled/disabled depending on what we are looking for in a particular run. An example is the prop_ we added in the naive rewriter. A test run or an executable run of oxide should be able ro enable/disable different properties at different levels. Consider it an error/warning/trace etc.

I have been looking into the existing tracing libraries for (2) but didn't manage to find one that does everything I think we need. I think we need named properties, registered centrally (maybe through codegen) and when at call site we decide how to interpret each property. (1) is this PR, (2) is a separate but related one...
```

niklasdewally commented:
> I  don't understand what do we want here...

A current example is our check for multiple applicable rules. This is currently done as an argument passed in, but when we have a lot of them managing them in our context object / using a library might be better.

See #609 #591

I will read the rest later :)

niklasdewally commented:
@YehorBoiar overall lgtm.

In case you didn't see this -  https://github.com/conjure-cp/conjure-oxide/issues/523#issuecomment-2572971153.


We could potentially benefit from writing our own a custom test runner so that we don't have to keep hacking around rusts one.

YehorBoiar commented on conjure_oxide/tests/generated_tests.rs:
```diff
@@ -42,10 +43,18 @@ use pretty_assertions::assert_eq;
+    enable_2b: Option<bool>,
+    disable_3a: Option<bool>,
+    enable_3b: Option<bool>,
+    disable_4a: Option<bool>,
```
@niklasdewally @ozgurakgun I made it so we would need to explicitly disable main stages, and implicitly enable secondary stages. Does it make sence, or would it be a bit confusing for others?

niklasdewally commented:
It might be clearer to have names here instead of (or as well as) the numbers.

Also, for consistency, having all settings be opt-in (e.g. run_x, enable_x) instead of  a mixture of opt-in and opt-out would be better.

On 25 January 2025 13:47:40 GMT, Yehor Boiar ***@***.***> wrote:
>@YehorBoiar commented on this pull request.
>
>
>
>>  
>+    disable_1a: Option<bool>,
>+    enable_1b: Option<bool>,
>+    disable_2a: Option<bool>,
>+    enable_2b: Option<bool>,
>+    disable_3a: Option<bool>,
>+    enable_3b: Option<bool>,
>+    disable_4a: Option<bool>,
>
>@niklasdewally @ozgurakgun I made it so we would need to explicitly disable main stages, and implicitly enable secondary stages. Does it make sence, or would it be a bit confusing for others?
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/conjure-cp/conjure-oxide/pull/549#pullrequestreview-2574015185
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

YehorBoiar commented:
> It might be clearer to have names here instead of (or as well as) the numbers. Also, for consistency, having all settings be opt-in (e.g. run_x, enable_x) instead of a mixture of opt-in and opt-out would be better.
> […](#)
> On 25 January 2025 13:47:40 GMT, Yehor Boiar ***@***.***> wrote: @YehorBoiar commented on this pull request. > + disable_1a: Option<bool>, + enable_1b: Option<bool>, + disable_2a: Option<bool>, + enable_2b: Option<bool>, + disable_3a: Option<bool>, + enable_3b: Option<bool>, + disable_4a: Option<bool>, @niklasdewally @ozgurakgun I made it so we would need to explicitly disable main stages, and implicitly enable secondary stages. Does it make sence, or would it be a bit confusing for others? -- Reply to this email directly or view it on GitHub: [#549 (review)](https://github.com/conjure-cp/conjure-oxide/pull/549#pullrequestreview-2574015185) You are receiving this because you were mentioned. Message ID: ***@***.***>

Yeah, but then we would compare solutions against conjure by default, which is not very nice.

niklasdewally commented:
> > It might be clearer to have names here instead of (or as well as) the numbers. Also, for consistency, having all settings be opt-in (e.g. run_x, enable_x) instead of a mixture of opt-in and opt-out would be better.
> > […](#)
> > On 25 January 2025 13:47:40 GMT, Yehor Boiar ***@***.***> wrote: @YehorBoiar commented on this pull request. > + disable_1a: Option<bool>, + enable_1b: Option<bool>, + disable_2a: Option<bool>, + enable_2b: Option<bool>, + disable_3a: Option<bool>, + enable_3b: Option<bool>, + disable_4a: Option<bool>, @niklasdewally @ozgurakgun I made it so we would need to explicitly disable main stages, and implicitly enable secondary stages. Does it make sence, or would it be a bit confusing for others? -- Reply to this email directly or view it on GitHub: [#549 (review)](https://github.com/conjure-cp/conjure-oxide/pull/549#pullrequestreview-2574015185) You are receiving this because you were mentioned. Message ID: ***@***.***>
> 
> Yeah, but then we would compare solutions against conjure by default, which is not very nice


We can have fields default to false but still be worded as opt in.

niklasdewally commented:
Is this branch updated to main? I cleaned the settings code up a bit since this pr was opened: primarily to do with defaults handling and setting names.

Also, I changed the types of the settings to be bool not Option<bool>

YehorBoiar commented:
> > > It might be clearer to have names here instead of (or as well as) the numbers. Also, for consistency, having all settings be opt-in (e.g. run_x, enable_x) instead of a mixture of opt-in and opt-out would be better.
> > > […](#)
> > > On 25 January 2025 13:47:40 GMT, Yehor Boiar _**@**_._**> wrote: @YehorBoiar commented on this pull request. > + disable_1a: Option, + enable_1b: Option, + disable_2a: Option, + enable_2b: Option, + disable_3a: Option, + enable_3b: Option, + disable_4a: Option, @niklasdewally @ozgurakgun I made it so we would need to explicitly disable main stages, and implicitly enable secondary stages. Does it make sence, or would it be a bit confusing for others? -- Reply to this email directly or view it on GitHub: [#549 (review)](https://github.com/conjure-cp/conjure-oxide/pull/549#pullrequestreview-2574015185) You are receiving this because you were mentioned. Message ID: _**@**_.**_>
> > 
> > 
> > Yeah, but then we would compare solutions against conjure by default, which is not very nice
> 
> We can have fields default to false but still be worded as opt in.

Like conjure_check: Option<bool> = false;?

niklasdewally commented:
I would prefer verb_noun: e.g enable_conjure_checks, check_solutions_against_conjure


Also see

https://github.com/conjure-cp/conjure-oxide/pull/560

YehorBoiar commented:
I see! I should probably rebase on main branch in this case

niklasdewally commented:
> I see! I should probably rebase on main branch in this case 

I'm inclined to think that we should add these changes as part of a larger rewrite of the integration tester to use our own test harness, as linked above.

There's a few other bugbears and  desired features that will be hard to hack onto the existing system, and a rewrite might be needed.

See the issues linked above and the PR linked in them. 

Might be worth holding back on the his until we discuss this in a group meeting with Soph and Liz as well, and gather a more complete list of requirements for a refactor rewrite.


Cc @ozgurakgun

niklasdewally commented:
In particular, Oz wanted cli arguments for each config argument, and that might make this PR redundant.

YehorBoiar commented:
@ozgurakgun Can you help me figuring out what other stuff I should implement in this PR?

YehorBoiar commented:
@ozgurakgun Do we want to have access to TestConfig variables in CLI?

ozgurakgun commented:
> @ozgurakgun Do we want to have access to TestConfig variables in command line?

Do you mean in the tester or the main executable? I'd say yes for the former and no for the latter.

YehorBoiar commented:
> > @ozgurakgun Do we want to have access to TestConfig variables in command line?
> 
> Do you mean in the tester or the main executable? I'd say yes for the former and no for the latter.

What are tester and main executables? The question was whether we want to be able to write something like `cargo test -- solve_with_minion=false` when running tests

niklasdewally commented on conjure_oxide/tests/generated_tests.rs:
```diff
@@ -150,210 +234,264 @@ fn integration_test_inner(
+        None
+    };
+
+    // Stage 3b: Check solutions against Conjure (only if explicitly enabled)
+    if config.compare_solver_solutions {
+        let conjure_solutions: Vec<BTreeMap<Name, Literal>> =
+            get_solutions_from_conjure(&format!("{}/{}.{}", path, essence_base, extension))?;
+
+        let username_solutions = normalize_solutions_for_comparison(
+            solutions.as_ref().expect("Minion solutions required"),
+        );
+        let conjure_solutions = normalize_solutions_for_comparison(&conjure_solutions);
+
+        let mut conjure_solutions_json = minion_solutions_to_json(&conjure_solutions);
+        let mut username_solutions_json = minion_solutions_to_json(&username_solutions);
+
+        conjure_solutions_json.sort_all_objects();
+        username_solutions_json.sort_all_objects();
+
+        assert_eq!(
+            username_solutions_json, conjure_solutions_json,
+            "Solutions do not match conjure!"
+        );
```
Is this enabled by default when ACCEPT=true?

YehorBoiar replied:
Forgot to add it. Good catch

YehorBoiar replied:
Ah, I did separate it on purpose. I don't think accept should run conjure, so I separated accept block with compare_solver_solutions. However, I'm open to discuss whether we should leave it as it was or use my solution @ozgurakgun @niklasdewally

ozgurakgun commented:
@YehorBoiar this needs to be brought up to date with main before I can merge.

YehorBoiar commented:
It looks like it doesn't like parsed models I gave it

niklasdewally commented:
> It looks like it doesn't like parsed models I gave it

When comparing JSON to JSON, you'll need to ignore the id fields, as these are not guaranteed to be the same across runs.

The Eq/PartialEq implementations for SymbolTable/ Model / etc. already id fields so work as expected.

ozgurakgun commented:
Did you run the tests with ACCEPT @YehorBoiar? I expected that to work as well, though what @niklasdewally says would mean we would have to update the JSON files less frequently.

YehorBoiar commented:
When accept=true we run minion solutions (because we want to compare them against conjure). In the case of optimizations tests - we don't want to run minion and conjure (as it would require too much time to solve the problem). What should I do? Should I not check against conjure if solve_with_minion is false?

ozgurakgun commented:
For this PR, just run Minion? The general question, I guess we need to have the option to do either.

YehorBoiar commented:
For some reason it complains on parser decpite me running it with ACCEPT flag :/ 

I compared solutions manually, and they are identical in both files. I think I would need a hand closing this PR. Can we do it on the upcoming meeting? @ozgurakgun

ozgurakgun commented:
@niklasdewally - are the IDs stable across reruns, assuming the code and the input are identical? What do they depend on?

niklasdewally commented:
> @niklasdewally - are the IDs stable across reruns, assuming the code and the input are identical? What do they depend on?

Ids are not stable. They are thread-local variables that depend on the creation order of symbol tables in that thread. I think cargo test schedules multiple tests per thread non-deterministically, so these do change from run to run.

I have defined `Eq` for SymbolTables to ignore the id field, so if you are comparing two `Model` objects, the ids changing shouldn't matter.


If the only thing that changes between the expected and generated models are the ids of the symbol tables, the models should still be equal, so the tests should pass without `ACCEPT=true`.

niklasdewally commented:
Assuming that we are still turning the JSON into models and comparing them as models, not directly comparing the JSON itself.

ozgurakgun commented:
What @YehorBoiar was observing earlier today was the following. He has a model object that is the result of rewriting. An expected model object which is read from the expected file. There are not equal. If instead he reads the generated and expected models from the corresponding files, they are equal. So currently in this PR he reads expected and generated both from files, converts them to Model and compares. I don't know why the Model object (indeed this is what he writes to both of those files if accept=true) does not compare as equal.

niklasdewally commented:
I have no idea...

There are also ids in declarations: do we know if it was declarations or symbol tables causing the problem?

From the diffs, it looks like it is the declaration ids changed in this PR but not the symbol table ones?

Declaration also ignores id for `Eq`, so this also shouldn't affect things too much...

niklasdewally commented:
Could I have a minimal reproducible example for this?  I'll be happy to run this through a debugger, but I'm not too familiar with the contents of this PR and what the problematic lines are.

ozgurakgun commented on conjure_oxide/tests/generated_tests.rs:
```diff
@@ -150,210 +222,266 @@ fn integration_test_inner(
+    // Check Stage 1a (parsed model)
+    if config.parse_model_default {
+        let expected_model = read_model_json(&context, path, essence_base, "expected", "parse")?;
+        let model_from_file = read_model_json(&context, path, essence_base, "generated", "parse")?;
```
@niklasdewally: around here instead of model_from_file, using model.expect("") should recreate the issue. 

@YehorBoiar: is that right?

niklasdewally replied:
Was there a particular parser that this was broken with?

ozgurakgun replied:
You mean test case? @YehorBoiar will have to remember / dig it up, I don't remember which one it was, sorry.

niklasdewally replied:
Is this bug specific to the treesitter parser or the json parser?

niklasdewally replied:
(if thats unknown, I'll try both)

ozgurakgun replied:
I don't know.

niklasdewally replied:
Should model.clone() deep clone or shallow clone a model?

I think the symbol tables are staying linked on clone, making machine names appear in the parsed model output...

niklasdewally commented:
@YehorBoiar didn't realise you weren't on the latest commit of main - the merge you just did might've pulled in some stuff that fixes your expects issue?

YehorBoiar commented:
@ozgurakgun Merge?

YehorBoiar commented:
> @YehorBoiar didn't realise you weren't on the latest commit of main - the merge you just did might've pulled in some stuff that fixes your parsed model issue?
> 
> (The commits you pulled in did quite a lot of changes to model)

Btw, I did double checked it. It didn't fix the issue

ozgurakgun commented:
Thanks @YehorBoiar!
